voxblox/include/voxblox/integrator/integrator_utils.h

```diff
@@ -12,25 +12,121 @@
+// also accesses neighboring indexes through thread safe locks, the amount of
+// time a thread must wait on a lock is minimized.
+class ThreadSafeIndex {
+ public:
+  ThreadSafeIndex(const size_t number_of_points, const size_t number_of_threads)
+      : number_of_points_(number_of_points),
+        number_of_threads_(number_of_threads),
+        atomic_idx_(0) {}
+
+  // returns true if index is valid, false otherwise
+  bool getNextIndex(size_t* idx) {
+    DCHECK(idx != nullptr);
+    size_t sequential_idx = atomic_idx_.fetch_add(1);
+
+    if (sequential_idx >= number_of_points_) {
+      return false;
+    } else {
+      *idx = getMixedIndex(sequential_idx);
+      return true;
+    }
+  }
+
+  void reset() { atomic_idx_.store(0); }
+
+ private:
+  size_t getMixedIndex(size_t base_idx) {
+    const size_t number_of_groups = number_of_threads_;
+    const size_t points_per_group = number_of_points_ / number_of_groups;
+
+    if (number_of_groups * points_per_group <= base_idx) {
+      return base_idx;
+    }
+
+    const size_t group_num = base_idx % number_of_groups;
+    const size_t position_in_group = base_idx / number_of_groups;
+
+    return group_num * points_per_group + position_in_group;
+  }
+
+  const size_t number_of_points_;
+  const size_t number_of_threads_;
+  std::atomic<size_t> atomic_idx_;
+};
+
+// This class assumes PRE-SCALED coordinates, where one unit = one voxel size.
+// The indices are also returned in this scales coordinate system, which should
+// map to voxel indices.
 class RayCaster {
  public:
+  RayCaster(const Point& origin, const Point& point_G,
+            const bool is_clearing_ray, const bool voxel_carving_enabled,
+            const FloatingPoint max_ray_length_m,
+            const FloatingPoint voxel_size_inv,
+            const FloatingPoint truncation_distance,
+            const bool cast_from_origin = true) {
+    const Ray unit_ray = (point_G - origin).normalized();
+
+    const Point ray_end = is_clearing_ray
+                              ? origin + unit_ray * max_ray_length_m
+                              : point_G + unit_ray * truncation_distance;
+    const Point ray_start = voxel_carving_enabled
+                                ? origin
+                                : (point_G - unit_ray * truncation_distance);
+
+    const Point start_scaled = ray_start * voxel_size_inv;
+    const Point end_scaled = ray_end * voxel_size_inv;
+
+    if (cast_from_origin) {
+      setupRayCaster(start_scaled, end_scaled);
+    } else {
+      setupRayCaster(end_scaled, start_scaled);
+    }
+  }
+
   RayCaster(const Point& start_scaled, const Point& end_scaled) {
+    setupRayCaster(start_scaled, end_scaled);
+  }
+
+  // returns false if ray terminates at ray_index, true otherwise
+  bool nextRayIndex(AnyIndex* ray_index) {
+    if (at_end_) {
+      return false;
+    }
+    DCHECK(ray_index != nullptr);
+    *ray_index = curr_index_;
+    at_end_ = curr_index_ == end_index_;
```

Comment:
yeah, it was finishing one too early by immediately returning at_end when it should return it on the next call.